azuredevops/internal/service/graph/data_service_principal.go

```diff
@@ -0,0 +1,162 @@
+		}
+		return fmt.Errorf("%s. Error: %v", errMsg, err)
+	}
+
+	d.SetId(*servicePrincipal.Descriptor)
+	d.Set("descriptor", servicePrincipal.Descriptor)
+	d.Set("display_name", servicePrincipal.DisplayName)
+	d.Set("origin_id", servicePrincipal.OriginId)
+	d.Set("origin", servicePrincipal.Origin)
+
+	/*
+		storageKey, err := clients.GraphClient.GetStorageKey(clients.Ctx, graph.GetStorageKeyArgs{
+			SubjectDescriptor: servicePrincipal.Descriptor,
+		})
+		if err != nil {
+			return err
+		}
+
+		if storageKey.Value != nil {
+			d.Set("service_principal_id", storageKey.Value.String())
+		}
+	*/
+	return nil
+}
+
+// Query AZDO for service principals with matching filter and search string
+func getIdentityServicePrincipalsWithFilterValue(clients *client.AggregatedClient, searchFilter *string, filterValue string) (*[]identity.Identity, error) {
+	// Get list of user with search filter and filter value provided at data source invocation.
+	response, err := clients.IdentityClient.ReadIdentities(clients.Ctx, identity.ReadIdentitiesArgs{
+		SearchFilter: searchFilter, // Filter to get users
+		FilterValue:  &filterValue,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	return response, nil
+}
+
+// Flatten Query Results
+func flattenIdentityServicePrincipals(servicePrincipals *[]identity.Identity) (*[]identity.Identity, error) {
+	if servicePrincipals == nil {
+		return nil, fmt.Errorf(" Input Service Principals Parameter is nil")
+	}
+	results := make([]identity.Identity, len(*servicePrincipals))
+	for i, servicePrincipal := range *servicePrincipals {
+		if servicePrincipal.Descriptor == nil {
+			return nil, fmt.Errorf(" User Object does not contain an id")
+		}
+		newUser := identity.Identity{
+			Id:                  servicePrincipal.Id,
+			Descriptor:          servicePrincipal.Descriptor,
+			ProviderDisplayName: servicePrincipal.ProviderDisplayName,
+			SubjectDescriptor:   servicePrincipal.SubjectDescriptor,
+			// Add other fields here if needed
+		}
+		results[i] = newUser
+	}
+	return &results, nil
+}
+
+// Filter results to validate user is correct. Occurs post-flatten due to missing properties based on search-filter.
+func validateIdentityServicePrincipal(servicePrincipals *[]identity.Identity, displayName string) *identity.Identity {
+	for _, servicePrincipal := range *servicePrincipals {
+		if strings.Contains(strings.ToLower(*servicePrincipal.ProviderDisplayName), strings.ToLower(displayName)) {
+			return &servicePrincipal
+		}
+	}
+	return nil
+}
+
+func getServicePrincipal(clients *client.AggregatedClient, servicePrincipalDescriptor *string) (*graph.GraphServicePrincipal, error) {
+	args := graph.GetServicePrincipalArgs{}
+	if servicePrincipalDescriptor != nil {
+		args.ServicePrincipalDescriptor = servicePrincipalDescriptor
+	}
+	response, err := clients.GraphClient.GetServicePrincipal(clients.Ctx, args)
+	if err != nil {
+		return nil, err
+	}
+	return response, nil
```

Comment:
No need to check the `descriptor` as this is the only parameter needed
```suggestion
	return clients.GraphClient.GetServicePrincipal(clients.Ctx, graph.GetServicePrincipalArgs{
		ServicePrincipalDescriptor: servicePrincipalDescriptor,
	})
```